 may throw an exception; if it returns normally, the caller
  // will need to conti